Objects/longobject.c

```diff
@@ -3676,7 +3676,17 @@ long_hash(PyObject *obj)
     }
     i = _PyLong_DigitCount(v);
     sign = _PyLong_NonCompactSign(v);
-    x = 0;
+
+    // unroll first two digits
+    assert(i>=2);
+    --i;
+    x = v->long_value.ob_digit[i];
+    assert(x < _PyHASH_MODULUS);
+    --i;
+    x = ((x << PyLong_SHIFT));
+    x += v->long_value.ob_digit[i];
+    assert(x < _PyHASH_MODULUS);
```

Comment:
```suggestion
    if (x >= _PyHASH_MODULUS)
        x -= _PyHASH_MODULUS;
```
This assert will trigger for values like `hash(2**31)` for 32bit targets which use the default `#define PYLONG_BITS_IN_DIGIT 30`
https://github.com/python/cpython/blob/0d4fd10fbab2767fad3eb27639905c8885b88c89/Include/pyport.h#L132-L138
because
https://github.com/python/cpython/blob/0d4fd10fbab2767fad3eb27639905c8885b88c89/Include/pyport.h#L170
and thus `PyHASH_BITS` is 31
https://github.com/python/cpython/blob/0d4fd10fbab2767fad3eb27639905c8885b88c89/Include/cpython/pyhash.h#L12-L18

I suggest to add more tests to `test_long_hash` around these corner cases for such build configurations:
```
>>> hash(-2**31 - 2)
-3
>>> hash(-2**31 - 1)
-2
>>> hash(-2**31)
-2
>>> hash(2**31)
1
>>> hash(2**31 + 1)
2
>>> hash(2**31 + 2)
3
```

Reply:
I added a check that `hash(-2**31)` is not `-1`. For the others I has a bit hesitant as the values are implementation details (document as implementation details here: https://docs.python.org/3/library/stdtypes.html#hashing-of-numeric-types). To add tests I would have to get the (private) value of `PyHASH_MODULUS`.

Reply:
Interesting, your link also mentions that `PyHASH_MODULUS` is available via https://docs.python.org/3/library/sys.html#sys.hash_info.modulus, which is e.g. used in
https://github.com/python/cpython/blob/283b05052338dd735cd4927011afc3735d9c6c7c/Lib/test/test_numeric_tower.py#L14
and some other tests. But yeah, seems to be an implementation detail. Don't know for sure whether it's ok to use it in tests - but I think so?

Reply:
PyHASH_MODULUS is not private anymore.  And it was documented in the algorithm description before.

Reply:
The "implementation details" are fully documented and `PyHASH_MODULUS` is indeed public. So I agree we need tests for the corner cases. I rewrote the tests to use `sys.hash_info.modulus`.

In `test_numeric_tower.py` there are more tests for the hash of ints, but these tests added here in this PR are not included there directly (only indirectly, e.g. by comparing the value of `hash(n)` with `hash(float(n))`.